### PR TITLE
Expired beacons fix.

### PR DESCRIPTION
**Pull request by Aadeesh11**
Fixes: #20 
Expected Behaviour: 

- [X] Show expired beacons at the very last in your beacons section.
- [X] Cancel subscriptions (if subscriptions have already been setup) and not setup them up at all (if user opens an expired  beacon).
- [X] Show snackBar if user tries to join expired beacon.
- [x] Hide the share button if beacon has expired. 
- [ ] Stop user from joining expired beacon (Would be handled by backend)

### CODE REVIEW

**Comment by Aadeesh11** [comment]
Should the last(4th) be part of backend logic or frontend? Because if it is a part of frontend logic, then while joining we would need an extra network call to get beacon data(since we only know shortcode at the time of joining).
@nb9960  please let me know on how to proceed from here.

**Comment by cfsmp3** [comment]
> Should the last(5th) be part of backend logic or frontend? 

Anything related to security / privacy / etc must be handled in the backend - you just can't trust the frontend :-)

So the backend should prevent joining an expired beacon, return some error, and then the frontend can capture that error and display it.

**Comment by Aadeesh11** [comment]
> > Should the last(5th) be part of backend logic or frontend?
> 
> Anything related to security / privacy / etc must be handled in the backend - you just can't trust the frontend :-)
> 
> So the backend should prevent joining an expired beacon, return some error, and then the frontend can capture that error and display it.

Ok understood, would create a new issue regarding this in the beacon-backend repository.

**Comment by Aadeesh11** [comment]
CI would pass once #42 gets merged.

File: lib/views/hike_screen.dart [hunk 1]
```diff
@@ -126,7 +129,15 @@ class _HikeScreenState extends State<HikeScreen> {
...
     }
-    final mergeStreamSubscription = mergedStream.listen((event) async {
+    StreamSubscription<dynamic> mergeStreamSubscription;
+    mergeStreamSubscription = mergedStream.listen((event) async {
+      if (DateTime.fromMillisecondsSinceEpoch(beacon.expiresAt)
+          .isBefore(DateTime.now())) {
+        setState(() {
+          isBeaconExpired = true;
+        });
+        mergeStreamSubscription.cancel();
```
Comment (nb9960) [inline lib/views/hike_screen.dart:139]:
```suggestion
        mergeStreamSubscription.cancel();
        setState(() {
          isBeaconExpired = true;
        });
```

**Review by nb9960 — CHANGES_REQUESTED** [review]

**Comment by nb9960** [comment]
Also, can you change the beacon exit dialog when the beacon has expired? Currently, the message is: 'There are ${X - 1} followers and you are carrying the beacon. Do you want to terminate the hike?'. Change it to 'Are you sure you want to exit? or something similar' **if** beacon has expired.

**Comment by Aadeesh11** [comment]
> Also, can you change the beacon exit dialog when the beacon has expired? Currently, the message is: 'There are ${X - 1} followers and you are carrying the beacon. Do you want to terminate the hike?'. Change it to 'Are you sure you want to exit? or something similar' **if** beacon has expired.

Would change that too.

**Review by nb9960 — APPROVED** [review]

**Comment by nb9960** [comment]
Thanks for the PR, merged! :tada:

**Pull request merged**